pyat/at/physics/radiation.py

```diff
@@ -372,5 +374,54 @@ def gen_quantdiff_elem(ring, orbit=None):
     return diff_elem
 
 
+@check_radiation(True)
+def tapering(ring, quadrupole=True, sextupole=True, niter=1,
+             **kwargs):
+    """
+    Scales magnet strength with local energy to cancel the closed orbit
+    and optics errors due to synchrotron radiations. PolynomB is used for
+    dipoles such that the machine geometry is maintained.
+    tapering(ring) or ring.tapering()
+    PARAMETERS
+        ring            lattice description.
+
+    KEYWORDS
+        qadrupole=True  scale quadrupole fields
+        sextupole=True  scale sextupole fields
+        niter=1         number of iteration
+        XYStep=1.0e-8   transverse step for numerical computation
+        DPStep=1.0E-6   momentum deviation used for computation of orbit6
+    """
+
+    xy_step = kwargs.pop('XYStep', DConstant.XYStep)
+    dp_step = kwargs.pop('DPStep', DConstant.DPStep)
+    dipoles = get_refpts(ring, Dipole)
+    b0 = get_value_refpts(ring, dipoles, 'BendingAngle')
+    ld = get_value_refpts(ring, dipoles, 'Length')
+
+    for i in range(niter):
+        o0, _ = find_orbit6(ring, XYStep=xy_step, DPStep=dp_step)
+        o6 = numpy.squeeze(lattice_pass(ring, o0, refpts=range(len(ring))))
+        dpps = (o6[4, dipoles] + o6[4, dipoles+1]) / 2
```

Comment:
`o6[4, dipoles+1]` will fail if the lattice ends with a dipole. You should use `range(len(ring)+1)` to compute the orbit.

Reply:
You are right, I have included the proposed changes to dipoles/qudrupoles and sextupoles

Concerning the loop, I had the same questions but it turns out that only one iteration is needed for quadrupoles and sextupoles. This is explained by the fact that when the closed orbit is minimized by the dipole loop, multipoles do not contribute significantly to energy loss anymore. In order the speed-up the calculation I have moved them out of the loop.